File: LeetCode/src/designclass/RandomizedSet.java
package designclass;

import java.util.*;

public class RandomizedSet {
    private Map<Integer, Integer> map;
    private List<Integer> list;
    private Random random;

    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
        random = new Random();
    }

    public boolean insert(int val) {
        if (map.containsKey(val)) {
            return false;
        } else {
            map.put(val, list.size());
            list.add(val);
            return true;
        }
    }

    public boolean remove(int val) {
        if (map.containsKey(val)) {
            int index = map.get(val);
            int lastElement = list.get(list.size() - 1);
            list.set(index, lastElement);
            map.put(lastElement, index);
            list.remove(list.size() - 1);
            map.remove(val);
            return true;
        } else {
            return false;
        }
    }

    public int getRandom() {
        return list.get(random.nextInt(list.size()));
    }

    public static void main(String[] args) {
        RandomizedSet solution = new RandomizedSet();

        System.out.println(solution.insert(1)); // true
        System.out.println(solution.remove(2)); // false
        System.out.println(solution.insert(2)); // true
        System.out.println(solution.getRandom()); // 1 or 2
        System.out.println(solution.remove(1)); // true
        System.out.println(solution.insert(2)); // false
        System.out.println(solution.getRandom()); // 2
    }
}

File: LeetCode_Python/designclass/RandomizedSet.py
import random


class RandomizedSet:
    def __init__(self):
        self.list = []
        self.dict = {}

    def insert(self, val: int) -> bool:
        if val in self.dict:
            return False

        self.dict[val] = len(self.list)
        self.list.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.dict:
            return False

        index = self.dict[val]
        last = self.list[-1]
        self.list[index] = last
        self.dict[last] = index
        self.list.pop()
        del self.dict[val]

        return True

    def getRandom(self) -> int:
        return random.choice(self.list)


if __name__ == "__main__":
    solution = RandomizedSet()
    print("Excepetd Output : True")
    print("Actual Output : ", solution.insert(1))
    print("Excepetd Output : False")
    print("Actual Output : ", solution.remove(2))
    print("Excepetd Output : True")
    print("Actual Output : ", solution.insert(2))
    print("Excepetd Output : 1 or 2")
    print("Actual Output : ", solution.getRandom())
    print("Excepetd Output : True")
    print("Actual Output : ", solution.remove(1))
    print("Excepetd Output : False")
    print("Actual Output : ", solution.insert(2))
